test/integration/garbagecollector/garbage_collector_test.go

```diff
@@ -1276,3 +1288,91 @@ func testCRDDeletion(t *testing.T, ctx *testContext, ns *v1.Namespace, definitio
+
+	// Create a fake apiservice to simulate discovery failed
+	apiservice, err := apiRegClient.APIServices().Create(context.TODO(), &apiregistration.APIService{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "APIService",
+			APIVersion: "apiregistration.k8s.io/v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "v1.test.io",
+		},
+		Spec: apiregistration.APIServiceSpec{
+			Service: &apiregistration.ServiceReference{
+				Namespace: "default",
+				Name:      "test",
+			},
+			Group:                "test.io",
+			Version:              "v1",
+			GroupPriorityMinimum: 1,
+			VersionPriority:      1,
+		},
+	}, metav1.CreateOptions{})
+	if err != nil {
+		t.Fatalf("Failed to create apiservice: %v", err)
+	}
+	defer apiRegClient.APIServices().Delete(context.TODO(), apiservice.Name, metav1.DeleteOptions{})
+
+	ns := createNamespaceOrDie("gc-discovery-fail", clientSet, t)
+	defer deleteNamespaceOrDie(ns.Name, clientSet, t)
+
+	testBlockingOwnerRefDoesBlock(t, ctx, ns, 1*time.Second)
+}
+
+func TestCacheSyncFailedOnGCStart(t *testing.T) {
+	ctx := setup(t, 0)
+	defer ctx.tearDown()
+	apiExtensionClient, dynamicClient, clientSet := ctx.apiExtensionClient, ctx.dynamicClient, ctx.clientSet
+
+	ns := createNamespaceOrDie("gc-cache-sync-fail", clientSet, t)
+	defer deleteNamespaceOrDie(ns.Name, clientSet, t)
+
+	// Create a def with `v1beta1` version and a cr of the def, then update the def added
+	// with "v1" version as default and a fake conversion webhook, it will cause the
+	// cache syncing failed.
+	def, dc := createRandomCustomResourceDefinition(t, apiExtensionClient, dynamicClient, ns.Name)
+	_, err := dc.Create(context.TODO(), newCRDInstance(def, ns.Name, names.SimpleNameGenerator.GenerateName("foo-")), metav1.CreateOptions{})
+	if err != nil {
+		t.Fatalf("Failed to create cr: %v", err)
+	}
+	def, err = apiExtensionClient.ApiextensionsV1().CustomResourceDefinitions().Get(context.TODO(), def.Name, metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("Failed to get cr: %v", err)
+	}
+	def.Spec.Conversion = &apiextensionsv1.CustomResourceConversion{
+		Strategy: apiextensionsv1.WebhookConverter,
+		Webhook: &apiextensionsv1.WebhookConversion{
+			ClientConfig: &apiextensionsv1.WebhookClientConfig{
+				Service: &apiextensionsv1.ServiceReference{
+					Namespace: ns.Name,
+					Name:      "foo",
+				},
+				CABundle: []byte("xo0o2e-="),
+			},
+			ConversionReviewVersions: []string{
+				"v1beta1", "v1",
+			},
+		},
+	}
+	defv1 := def.Spec.Versions[0].DeepCopy()
+	defv1.Name = "v1"
+	defv1beta1 := def.Spec.Versions[0]
+	defv1beta1.Served = false
+	defv1beta1.Storage = false
+	def.Spec.Versions = []apiextensionsv1.CustomResourceDefinitionVersion{
+		*defv1,
+		defv1beta1,
+	}
+	_, err = apiExtensionClient.ApiextensionsV1().CustomResourceDefinitions().Update(context.TODO(), def, metav1.UpdateOptions{})
+	if err != nil {
+		t.Fatalf("update def error: %v", err)
+	}
+
+	testBlockingOwnerRefDoesBlock(t, ctx, ns, 1*time.Second)
```

Comment:
same comment about polling for whatever we're waiting for before calling testBlockingOwnerRefDoesBlock rather than plumbing a one second delay down into that function